src/splice.test-d.ts

```diff
@@ -1,21 +1,179 @@
+import { pipe } from "./pipe";
 import { splice } from "./splice";
 
-it("reflects the type of `items` in the return value", () => {
-  const items: Array<number> = [];
-  const result = splice(items, 0, 0, []);
-  expectTypeOf(result).toEqualTypeOf<Array<number>>();
+describe("arrays", () => {
+  it("reflects the type of `items` in the return value", () => {
+    expectTypeOf(splice([] as Array<number>, 0, 0, [])).toEqualTypeOf<
+      Array<number>
+    >();
+  });
+
+  it("reflects the type of `replacement` in the return value", () => {
+    expectTypeOf(splice([], 0, 0, [] as Array<number>)).toEqualTypeOf<
+      Array<number>
+    >();
+  });
+
+  it("reflects the type of `replacement` in the return value (data-last)", () => {
+    expectTypeOf(pipe([], splice(0, 0, [] as Array<number>))).toEqualTypeOf<
+      Array<number>
+    >();
+  });
+
+  it("does not work with replacement of different type", () => {
+    // @ts-expect-error - wrong `replacement` type
+    splice([1, 2, 3] as const, 0, 0, ["a"] as const);
+  });
 });
 
-it("reflects the type of `replacement` in the return value", () => {
-  const replacement: Array<number> = [];
-  const result = splice([], 0, 0, replacement);
-  expectTypeOf(result).toEqualTypeOf<Array<number>>();
+describe("fixed tuples", () => {
+  describe("deleteCount = 0", () => {
+    it("works with start + length < 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, -4, 0, [4] as const),
+      ).toEqualTypeOf<[4, 1, 2, 3]>();
+    });
+
+    it("works with start < 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, -1, 0, [4] as const),
+      ).toEqualTypeOf<[1, 2, 4, 3]>();
+    });
+
+    it("works with start = 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 0, 0, [4] as const),
+      ).toEqualTypeOf<[4, 1, 2, 3]>();
+    });
+
+    it("works with start > 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, 0, [4] as const),
+      ).toEqualTypeOf<[1, 4, 2, 3]>();
+    });
+
+    it("works with start = length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 3, 0, [4] as const),
+      ).toEqualTypeOf<[1, 2, 3, 4]>();
+    });
+
+    it("works with start > length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 6, 0, [4] as const),
+      ).toEqualTypeOf<[1, 2, 3, 4]>();
+    });
+  });
+
+  describe("deleteCount != 0", () => {
+    it("works with deleteCount < 0", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, -1, [4] as const),
+      ).toEqualTypeOf<[1, 4, 2, 3]>();
+    });
+
+    it("works with deleteCount + start < length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, 2, [4] as const),
+      ).toEqualTypeOf<[1, 4]>();
+    });
+
+    it("works with deleteCount + start > length", () => {
+      expectTypeOf(
+        splice([1, 2, 3] as const, 1, 4, [4] as const),
+      ).toEqualTypeOf<[1, 4]>();
+    });
```

Comment:
Let's add some values to the input array in the tests so that they expect different outputs. This will help protect against cases where a bug breaks one of them but hides it because the input is too simple.